src/draw/DefaultDrawer.js

```diff
@@ -106,466 +160,464 @@ class DefaultDrawer {
+      .data((data) => data[1])
+      .join('g')
+      .attr('id', (data) => data.data.id)
+      .on('click', clicked)
+      .call(drag)
+      .attr('transform', (data) => `translate(${data.x0},${data.y0})`);
+
+    node
+      .filter((data) => data.data.id !== '__shadowRoot')
+      .attr('class', (data) => `component
+        component-${data.data.definition.model}
+        ${data.data.definition.type}`)
+      .html((data) => this.resources.models[data.data.definition.model])
+      .select('svg')
+      .attr('id', (data) => `svg-${data.data.id}`)
+
+      .attr('height', (data) => data.y1 - data.y0)
+      .attr('width', (data) => data.x1 - data.x0);
+
+    node.select('.component-name')
+      .text((data) => data.data.name);
+    node.select('.component-type')
+      .text((data) => data.data.definition.type);
+    node.select('.component-icon')
+      .html((data) => this.resources.icons[data.data.definition.icon]);
+
+    node.select('rect')
+      .filter((d) => d.data?.definition?.isContainer)
+      .attr('height', (data) => data.y1 - data.y0)
+      .attr('width', (data) => data.x1 - data.x0);
+
+    node.select('.component-container')
+      .attr('height', (data) => Math.max(
+        data.y1
+            - (data.y0 + this.minHeight)
+            - this.margin,
+        this.minHeight,
+      ))
+      .attr('width', (data) => Math.max(data.x1 - data.x0 - 2 * this.margin, this.minWidth))
+      .attr('x', () => this.margin)
+      .attr('y', () => this.minHeight)
+      .filter((data) => data.children)
+      .append((data) => d3.select(`#group-${data.data.id}`).node());
+
+    node.select('.component-container>rect').attr('data-parentId', (d) => d.data.id);
+  }
+
+  /**
+   * Build d3 hierarchy and treemap layout.
+   * @return {[[groupId, [Node]]]} - The nodes grouped by parent.
+   */
+  buildTree() {
+    const treemapLayout = d3.treemap()
+      .size([this.width, this.height])
+      .tile((data) => {
+        const lines = this.__buildLines(data);
+
+        this.setupTiles(lines);
+        // TODO save/load coordinates
+      })
+      .round(true);
+
+    const rootNode = d3.hierarchy(this.shadowRoot);
+
+    rootNode
+      .count()
+      .sort((a, b) => b.height - a.height
+            || b.value - a.value
+            || (b.data.definition?.isContainer ?? 0)
+            - (a.data.definition?.isContainer ?? 0));
+
+    treemapLayout(rootNode);
+
+    return d3.groups(
+      rootNode,
+      (data) => (data.parent
+        && data.parent.data.id !== '__shadowRoot'
+        ? `group-${data.parent.data.id}`
+        : 'root-components'),
+    )
+      .filter((data) => data[0] !== 'root-__shadowRoot');
+  }
+
+  /**
+   * Get the anchor point on source selection facing target selection
+   * @param {Selection} sourceSelection - The source D3 selection object
+   * @param {Selection} targetSelection - The target D3 selection object
+   * @returns {[Number,Number] || null} - Tuple representing x,y coordinates,
```

Comment:
same here and to make or in type please use : `@return {Number|null}`
no s to return

Reply:
why don't you return an object with x and y property ? 
-> if you return directly object you can directly `return this.screenToSVG(anchorPoint.x, anchorPoint.y);`

Reply:
d3 link.source and link.target expect a Tuple representing x and y, hence the return format